[RHDHBUGS-2645]: Document Operator cache-level label filtering for large clusters (#2270)
redhat-developer/red-hat-developers-documentation-rhdh: modules/configure_configuring-rhdh/proc-optimize-operator-memory-usage-for-large-clusters.adoc

:_mod-docs-content-type: PROCEDURE

[id="optimize-operator-memory-usage-for-large-clusters_{context}"]
= Optimize Operator memory usage for large clusters

[role="_abstract"]
In large clusters, the {product} Operator can consume substantial memory because the cache loads metadata for all Secrets and ConfigMaps cluster-wide. You can enable cache-level label filtering to restrict caching to only relevant resources.

[IMPORTANT]
====
Enabling cache label filtering is a breaking change. After activation, every Secret and ConfigMap that the Operator manages must carry the `rhdh.redhat.com/external-config: "true"` label. Any existing {product-short} instances that reference unlabeled resources stop working until you add the required label to those resources.
====

.Prerequisites

* An {ocp-short} administrator has installed the {product} Operator.

> * You have identified the Secrets and ConfigMaps that your {product-short} instances reference.

Author: ```suggestion
* You have identified the secrets and config maps that your {product-short} instances reference.
```

.Procedure

. Enable cache-level label filtering by setting the `ENABLE_CACHE_LABEL_FILTER` environment variable on the Operator.
+
Edit the Operator `Subscription` resource and add the environment variable:
+
[source,yaml]
----
spec:
  config:
    env:
      - name: "ENABLE_CACHE_LABEL_FILTER"
        value: "true"
----
+

> Alternatively, you can edit the `ClusterServiceVersion` (CSV) resource to add the `--enable-cache-label-filter` command-line flag to the `manager` container:

Author: ```suggestion
Alternatively, edit the `ClusterServiceVersion` (CSV) resource to add the `--enable-cache-label-filter` command-line flag to the `manager` container:
```

+
[source,yaml]
----
spec:
  template:
    spec:
      containers:
      - name: manager
        args:
        - --enable-cache-label-filter
----

> . Label every Secret and ConfigMap that your {product-short} instances reference. This includes all resources listed in your `{product-custom-resource-type}` CR under `spec.application`, such as app-config ConfigMaps, secrets, and database credentials.

Author: ```suggestion
. Label every secret and config map that your {product-short} instances reference.
This includes all resources listed in your `{product-custom-resource-type}` custom resource under `spec.application`.
```

+
[source,terminal,subs="+quotes"]
----
$ oc label secret _<secret_name>_ rhdh.redhat.com/external-config=true -n _<namespace>_
$ oc label configmap _<configmap_name>_ rhdh.redhat.com/external-config=true -n _<namespace>_
----
+

> You can also add the label directly in the resource YAML:

Author: ```suggestion
Alternatively, add the label directly in the resource YAML:
```

:_mod-docs-content-type: PROCEDURE

[id="optimize-operator-memory-usage-for-large-clusters_{context}"]
= Optimize Operator memory usage for large clusters

[role="_abstract"]
In large clusters, the {product} Operator can consume substantial memory because the cache loads metadata for all Secrets and ConfigMaps cluster-wide. You can enable cache-level label filtering to restrict caching to only relevant resources.

[IMPORTANT]
====
Enabling cache label filtering is a breaking change. After activation, every Secret and ConfigMap that the Operator manages must carry the `rhdh.redhat.com/external-config: "true"` label. Any existing {product-short} instances that reference unlabeled resources stop working until you add the required label to those resources.
====

.Prerequisites

* An {ocp-short} administrator has installed the {product} Operator.
* You have identified the secrets and config maps that your {product-short} instances reference.

.Procedure

> 

Reviewer: Can lines 21 and 23 be combined? Maybe something like this:

To enable cache-level label filtering, edit the Operator Subscription resource and add the ENABLE_CACHE_LABEL_FILTER environment variable to the spec.config.env section:

Then show the YAML example

Reviewer: Also, PR needs rebase

. Enable cache-level label filtering by setting the `ENABLE_CACHE_LABEL_FILTER` environment variable on the Operator.
+

> Edit the Operator `Subscription` resource and add the environment variable:

Author: ```suggestion
. To enable cache-level label filtering, edit the Operator subscription and add the `ENABLE_CACHE_LABEL_FILTER` environment variable.
```
